src/economy/economy.cpp

```diff
@@ -519,64 +546,120 @@ float artisan_scale_limit(sys::state& state, dcon::nation_id n, dcon::commodity_
-			auto kf = state.world.commodity_get_key_factory(cid);
+	float total = 0.f;	
 
-			float market_size = state.world.commodity_get_total_consumption(cid) * 0.001f;
-			auto consumed_ratio = std::min(1.f, (state.world.commodity_get_total_consumption(cid) + 0.0001f) / (state.world.commodity_get_total_production(cid) + 0.0001f));
-			float output_size = state.world.commodity_get_artisan_output_amount(cid);
+	// crude approximation of softmax
+	for(uint32_t i = 1; i < scores.size(); ++i) {
+		distribution[i] = scores[i] - max_score;
+		distribution[i] = pseudo_exp_for_negative(distribution[i] * multiplier);
+		total += distribution[i];
+	}
 
-			if(state.world.commodity_get_artisan_output_amount(cid) > 0.0f && (state.world.commodity_get_is_available_from_start(cid) || (kf && state.world.nation_get_active_building(n, kf)))) {
-				auto& w = state.world.nation_get_artisan_distribution(n, cid);
-				if(profits[cid.index()] < 0) {
-					w = std::max(0.f, (1.0f + profits[cid.index()] * w * market_size / consumed_ratio / consumed_ratio / output_size - 100.f / num_artisans) * w);
-				} else if(total_weights > 0.0001f) {
-					float ideal_weight = (profits[cid.index()] / total_weights);
-					float speed = std::clamp(distribution_drift_speed * w * profits[cid.index()] * market_size / output_size + 100.f / num_artisans, 0.f, 0.01f);
-					w = std::clamp((1.0f - speed) * w + speed * ideal_weight, 0.f, 1.f);
-
-					assert(std::isfinite(w));
-				} else if (total_possible > 0.f) {
-					// move toward 0 production
-					w = (1.0f - distribution_drift_speed) * w;
-					assert(std::isfinite(w));
-				}
-			} else {
-				state.world.nation_set_artisan_distribution(n, cid, 0.0f);
-			}
+	for(uint32_t i = 1; i < scores.size(); ++i) {
+		distribution[i] = distribution[i] / (total + 0.001f);
+	}
+}
+
+void fill_artisan_distribution(sys::state& state, dcon::nation_id n, std::vector<float>& distribution) {
+	std::vector<float> scores;
+
+	auto const csize = state.world.commodity_size();
+
+	scores.resize(csize + 1);
+	distribution.resize(csize + 1);
+
+	for(uint32_t i = 1; i < csize; ++i) {
+		dcon::commodity_id cid{ dcon::commodity_id::value_base_t(i) };
+		scores[cid.index()] = state.world.nation_get_artisan_distribution(n, cid);
+	}
+
+	float mult = get_artisans_multiplier(state, n);
+	scores[csize] = 10.f / mult;
+
+	scores_to_distribution(scores, distribution, mult);
+}
+
+
+
+void adjust_artisan_balance(sys::state& state, dcon::nation_id n) {
+	auto const csize = state.world.commodity_size();
+	float distribution_drift_speed = 0.0001f;
+
+	std::vector<float> current_distribution;
+	std::vector<float> prev_profits;
+	std::vector<float> profits;
+	profits.resize(csize + 1);
+	current_distribution.resize(csize + 1);
+
+	float mult = get_artisans_multiplier(state, n);
+
+	profits[csize] = 10.f / mult; // * state.world.nation_get_everyday_needs_costs(n, state.culture_definitions.artisans) / state.defines.alice_needs_scaling_factor;
```

Comment:
I don't understand the intent of this. Here you initialize just the last value of the array. Did you mean to initialize all of them?

Reply:
No, this is to set the baseline for profit. If all profit (after adjustment) is small compared to this number, softmax will ensure that artisans will not produce goods if they are that unprofitable.